### ISSUE TITLE
posix_fadvise failing (under docker, on ARM)

### ISSUE THREAD

**Issue by ThomasWaldmann** [docker/docker#30516]
Please excuse me that I can't provide more infos about this myself, I only discovered this issue by accident by doing a web research, I didn't experience it myself.

I am one of the borg developers, but not a docker-on-ARM or Alpine Linux user.

**Description**

It seems there is an issue with the posix_fadvise call failing under docker on ARM platform:

https://bugs.alpinelinux.org/issues/6592

You can ignore the errno == 0 part of this issue, this seems to be an additional python wrapper bug.

But the posix_fadvise() call usually never fails, the first time we have seen this is in this bug report and we have that call since many months. It can't be due to the fileno passed to it as the same fileno is used in another call a few lines before and it did not fail there, so the fileno seems valid.

See also the discussion on the borg issue tracker:

https://github.com/borgbackup/borg/issues/2095

**Steps to reproduce the issue:**

See alpinelinux bug report.

**Describe the results you received:**

posix_fadvise() raising an OSError exception (with unknown errno, sadly). This means the call failed.

**Describe the results you expected:**

posix_fadvise() not raising an OSError. The call usually never fails.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

ARM platform.

borg has quite some ARM users, usually there is no problem caused by the ARM platform per se. So it seems to me that this is special to docker-on-ARM maybe.

**Comment by enkore** [comment]
(From Alpine Bug #6592)

```
Alpine Docker container on ARM:

Python 3.5.2 (default, Dec 20 2016, 17:55:24)
[GCC 6.2.1 20160822] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.posix_fadvise(0, 0, 0, os.POSIX_FADV_DONTNEED)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 0] Error
>>> 

Ubuntu 16.04 (ARM Host OS):

Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.posix_fadvise(0, 0, 0, os.POSIX_FADV_DONTNEED)
>>> 

Alpine Linux container on x86_64:

Python 3.5.2 (default, Dec 20 2016, 17:53:35) 
[GCC 6.2.1 20160822] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.posix_fadvise(0, 0, 0, os.POSIX_FADV_DONTNEED)
>>> 
```

**Comment by justincormack** [comment]
It could be a seccomp filtering issue on Arm, running with seccomp disabled
would test that.

Will take a look.

**Comment by justincormack** [comment]
This is seccomp related, as disabling makes it work. It fails oddly though.

**Comment by enkore** [comment]
The syscall behind posix_fadvise (fadvise64) is one of these that are different on ARM and ppc from x86, perhaps that's related to the cause

**Comment by justincormack** [comment]
By fails oddly, this test program:
```
#include <fcntl.h>

int main() {
	return posix_fadvise(0, 0, 0, POSIX_FADV_DONTNEED);
}
```
which runs correctly without seccomp:
```
/ # strace ./test
execve("./test", ["./test"], [/* 6 vars */]) = 0
set_tls(0xb6fd6548, 0xbe8b2cd0, 0x37, 0xbe8b2f6e, 0xb6fd64a0) = 0
set_tid_address(0xb6fd64bc)             = 19
mprotect(0xb6fd4000, 4096, PROT_READ)   = 0
mprotect(0x7f652000, 4096, PROT_READ)   = 0
fadvise64_64(0, 0, 0, POSIX_FADV_DONTNEED) = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```
fails like this by default:
```
/ # strace ./test
execve("./test", ["./test"], [/* 6 vars */]) = 0
set_tls(0x7f6320e4, 0xbc, 0x37, 0xbeac2f6e, 0x7f63203c) = 0
set_tid_address(0x7f632058)             = 58
exit_group(1)                           = ?
+++ exited with 1 +++
```
(if you believe `strace`).

**Comment by justincormack** [comment]
Ok, I think I know what this is, will prepare a patch.

**Comment by justincormack** [comment]
fixed by #30545

You can run `dockerd --seccomp-profile ...` with the profile downloaded from https://raw.githubusercontent.com/justincormack/docker/d6adcd6a82ba0997fcc123090711b5502da5cbfd/profiles/seccomp/default.json as a short term fix, will get this backported to 1.13.

**Comment by justincormack** [comment]
This will be in 1.13.1
